src/ssh_keydir/ssh_keydir.go

```diff
@@ -0,0 +1,400 @@
+	err := asUser(u, func() error {
+		td := stageDirPath(u)
+		if err := os.MkdirAll(td, 0700); err != nil {
+			return err
+		}
+		defer os.RemoveAll(td)
+
+		d, err := opendir(td)
+		if err != nil {
+			return err
+		}
+		d.User = u
+
+		akfb, err := ioutil.ReadFile(authKeysFilePath(u))
+		if err != nil && !os.IsNotExist(err) {
+			return err
+		} else if err == nil {
+			err = d.Add(PreservedKeysName, akfb, false, false)
+			if err != nil {
+				return err
+			}
+		}
+		if err = d.rename(authKeysDirPath(u)); err != nil {
+			return err
+		}
+		akd = d
+		return nil
+	})
+	return akd, err
+}
+
+// Open opens the authorized keys directory for the supplied user.
+// If create is false, Open will fail if no directory exists yet.
+// If create is true, Open will create the directory if it doesn't exist,
+// preserving the authorized_keys file in the process.
+// After a successful open, Close should be called when finished to unlock
+// the directory.
+func Open(usr string, create bool) (*SSHAuthorizedKeysDir, error) {
+	u, err := user.Lookup(usr)
+	if err != nil {
+		return nil, err
+	}
+
+	l, err := acquireLock(u)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			l.Close()
+		}
+	}()
+
+	akd, err := opendir(authKeysDirPath(u))
+	if err != nil && (!create || !os.IsNotExist(err)) {
+		return nil, err
+	} else if os.IsNotExist(err) {
+		akd, err = createAuthorizedKeysDir(u)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	akd.lock = l
+	akd.User = u
+	return akd, nil
+}
+
+// Close closes the authorized keys directory.
+func (akd *SSHAuthorizedKeysDir) Close() error {
+	return akd.lock.Close()
+}
+
+// rename renames the authorized_keys dir to the supplied path.
+func (akd *SSHAuthorizedKeysDir) rename(to string) error {
+	err := os.Rename(akd.Path, to)
+	if err != nil {
+		return err
+	}
+	akd.Path = to
+	return nil
+}
+
+// path returns the path to the named key.
+func (akd *SSHAuthorizedKeysDir) path(n string) string {
+	return filepath.Join(akd.Path, n)
+}
+
+// WalkKeys iterates across all keys in akd, calling f for each key.
+// Iterating stops on error, and the error is propagated out.
+func (akd *SSHAuthorizedKeysDir) WalkKeys(f func(*SSHAuthorizedKey) error) error {
+	d, err := os.Open(akd.Path)
+	if err != nil {
+		return err
+	}
+
+	names, err := d.Readdirnames(0)
+	if err != nil {
+		return err
+	}
+
+	sort.Strings(names)
+	for _, n := range names {
+		ak, err := akd.Open(n)
+		if err != nil {
+			return err
+		}
+		if err := f(ak); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Open opens the key at name.
+func (akd *SSHAuthorizedKeysDir) Open(name string) (*SSHAuthorizedKey, error) {
+	p := akd.path(name)
+	fi, err := os.Stat(p)
+	if err != nil {
+		return nil, err
+	}
+	ak := &SSHAuthorizedKey{
+		Name:     name,
+		Disabled: (fi.Size() == 0),
+		Path:     p,
+		origin:   akd,
+	}
+	return ak, nil
+}
+
+// Remove removes the opened key.
+func (ak *SSHAuthorizedKey) Remove() error {
```

Comment:
nit: My preference would be to group the SSHAuthorizedKeysDir and SSHAuthorizedKey functions together instead of grouping related functions of the two. Keeping the context of 'all these operate on this struct' while reading through linearly works better for me.